bitcoin/zcashd/zcashprices.go

```diff
@@ -0,0 +1,283 @@
+			return nil
+		}
+	}
+	log.Error("Failed to fetch zcash exchange rates")
+	return errors.New("All exchange rate API queries failed")
+}
+
+func (z *ZcashPriceFetcher) run() {
+	z.fetchCurrentRates()
+	ticker := time.NewTicker(time.Minute * 15)
+	for range ticker.C {
+		z.fetchCurrentRates()
+	}
+}
+
+func (provider *ExchangeRateProvider) fetch() (err error) {
+	if len(provider.fetchUrl) == 0 {
+		err = errors.New("Provider has no fetchUrl")
+		return err
+	}
+	resp, err := provider.client.Get(provider.fetchUrl)
+	if err != nil {
+		log.Error("Failed to fetch from "+provider.fetchUrl, err)
+		return err
+	}
+	decoder := json.NewDecoder(resp.Body)
+	var dataMap interface{}
+	err = decoder.Decode(&dataMap)
+	if err != nil {
+		log.Error("Failed to decode JSON from "+provider.fetchUrl, err)
+		return err
+	}
+	return provider.decoder.decode(dataMap, provider.cache, provider.bitcoinProvider)
+}
+
+func (b KrakenDecoder) decode(dat interface{}, cache map[string]float64, bp *exchange.BitcoinPriceFetcher) (err error) {
+	rates, err := bp.GetAllRates()
+	if err != nil {
+		return err
+	}
+	obj, ok := dat.(map[string]interface{})
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	result, ok := obj["result"]
+	if !ok {
+		return errors.New("KrakenDecoder: field `result` not found")
+	}
+	resultMap, ok := result.(map[string]interface{})
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	pair, ok := resultMap["BCHXBT"]
+	if !ok {
+		return errors.New("KrakenDecoder: field `BCHXBT` not found")
+	}
+	pairMap, ok := pair.(map[string]interface{})
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	c, ok := pairMap["c"]
+	if !ok {
+		return errors.New("KrakenDecoder: field `c` not found")
+	}
+	cList, ok := c.([]interface{})
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	rateStr, ok := cList[0].(string)
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	price, err := strconv.ParseFloat(rateStr, 64)
+	if err != nil {
+		return err
+	}
+	rate := price
+
+	if rate == 0 {
+		return errors.New("Bitcoin-BitcoinCash price data not available")
+	}
+	for k, v := range rates {
+		cache[k] = v * rate
+	}
+	return nil
+}
+
+func (b BitfinexDecoder) decode(dat interface{}, cache map[string]float64, bp *exchange.BitcoinPriceFetcher) (err error) {
+	rates, err := bp.GetAllRates()
+	if err != nil {
+		return err
+	}
+	obj, ok := dat.(map[string]interface{})
+	if !ok {
+		return errors.New("BitfinexDecoder type assertion failure")
+	}
+	r, ok := obj["last_price"]
+	if !ok {
+		return errors.New("BitfinexDecoder: field `last_price` not found")
+	}
+	rateStr, ok := r.(string)
+	if !ok {
+		return errors.New("BitfinexDecoder type assertion failure")
+	}
+	price, err := strconv.ParseFloat(rateStr, 64)
+	if err != nil {
+		return err
+	}
+	rate := price
+
+	if rate == 0 {
+		return errors.New("Bitcoin-BitcoinCash price data not available")
```

Comment:
Zcash again. This should be extracted to a package variable and shared across places it can occur.